### PR TITLE
Explain that dynamic refs can implement generics

### LINKED ISSUE #3601: Recommend describing templates/generic types using $dynamicRef
During the recent issue backlog tidying effort, I closed _at least_ six issues by pointing to @gregsdennis's [Using Dynamic References to Support Generic Types](https://json-schema.org/blog/posts/dynamicref-and-generics) blog post:

* #519 
* #555 
* #773 
* #957 
* #1749 
* #2453 

But there are several problems with pointing folks to this technique:

1. It's not mentioned in the OAS spec, so many people, including code generation tool developers, are not aware of it
2. It's not even mentioned in the JSON Schema spec as it wasn't how I was thinking about it when I wrote those sections
3. `$dynamicAnchor` and `$dynamicRef` are among the most likely features to be hiding in the asterisk on many tools' "Supports OAS 3.1*" claims (it's basically taken `discriminator`'s place as the keyword that is left until the end to even attempt)

We already have a large section on supporting polymorphic types.  We should add a brief section on generic types, at least enough to make it clear what keywords can be used.  We should definitely do this in 3.2.0, and maybe even 3.1.1 as it does not add any new requirements.  In 3.2.0, I suppose we could add requirements to support the idiom but I'd settle for making people aware of it.  If there's demand for it, tools will add support, but first folks have to know about it.

I don't think we need to go into a lot of detail on how it works.  That can be offloaded to the Learn site.

### PR DESCRIPTION
**Pull request by handrews**
Fixes #3601.

Includes an example.  This intentionally does not explain how dynamic referencing works, as there are better resources available in both the spec and (more readably) the official JSON Schema blog.

### CODE REVIEW

**Comment by vicspina** [comment]
Can you please take me off your list I don’t want no more emails from you
thank you

File: versions/3.1.1.md
Comment (ralfhandl) [inline versions/3.1.1.md:2353]:
```suggestion
* `$dynamicAnchor` identifies a set of possible schemas (including a default placeholder schema) to which a `$dynamicRef` can resolve
```

File: versions/3.1.1.md
Comment (ralfhandl) [inline versions/3.1.1.md:2351]: Could you add a link to some "intro" material? I couldn't find a good starting point in https://json-schema.org/draft/2020-12/json-schema-core.

Comment (handrews) [reply versions/3.1.1.md:2351]: @ralfhandl it's funny, I actually wrote a fairly enormous explanatory section and then cut it as I didn't want to bloat the section.  It's not trivial to explain.

The best explanation of this is @gregsdennis [post on the JSON Schema blog](https://json-schema.org/blog/posts/dynamicref-and-generics).  I'm not sure if we want to have a link to that in the OAS document?  Perhaps I could add a page to the learn site mentioning this and other JSON Schema techniques (including linking from the learn site to the JSON Schema blog?)

It's absolutely true that the JSON Schema spec is not clear on this, in part because we didn't realize `$dynamicRef` and `$dynamicAnchor` would support this when we came up with it.  Also, I have regrets about the writing in that whole section... many regrets...

Does the learn site idea sound good?  If so I will open a PR for that page and we can wait until we agree on it to merge this.  If not, I can try to add back the explanatory material here and we can figure out how much is worth including.  Or any combination of these approaches.

File: versions/3.1.1.md
Comment (gregsdennis) [inline versions/3.1.1.md:2770]: This may need to be in quotes. The syntax highlighting makes it look like a comment.

File: versions/3.1.1.md
Comment (gregsdennis) [inline versions/3.1.1.md:2354]:
```suggestion
* `$dynamicRef` resolves to the first matching `$dynamicAcnhor` encountered on its path from the schema entry point to the reference, as described in the JSON Schema specification
```

This has to be probably the most concise explanation of dynamic references that I've ever seen. 👍

File: versions/3.1.1.md
Comment (handrews) [reply versions/3.1.1.md:2770]: Oops, yeah that's a YAML comment, I'll fix, thanks!

File: versions/3.1.1.md
Comment (handrews) [reply versions/3.1.1.md:2354]: @gregsdennis it only took me three and a half years to figure out how to boil it down... 😅

Comment (handrews) [reply versions/3.1.1.md:2354]: Although you typo'd `$dynamicAcnhor` so I'll incorporate this separately as I don't know how to edit a suggestion.

Comment (ralfhandl) [reply versions/3.1.1.md]: Fixed proposal

File: versions/3.1.1.md
Comment (handrews) [reply versions/3.1.1.md:2351]: @ralfhandl so it turns out I already wrote a Learn page on dynamic references that links to the JSON Schema blog posts on generics.  I've updated the PR to link to that Learn page.  And also note up front that there is an example in the Schema Object Examples section as otherwise it seems a little abrupt.  Does that work for you?

Comment (ralfhandl) [reply versions/3.1.1.md:2351]: I'm a bit torn on directly linking the Learn page without mentioning it as an "Informative reference" in the appendix.

On the other hand there is precedence for direct linking already in v3.1.0, for example the links to https://yaml.org/spec/1.2/spec.html.

Comment (ralfhandl) [reply versions/3.1.1.md:2351]: That means we only have to figure out where in the build process this gets added, and ideally also produce an appendix for "Informative references" without changing the spec sources.

Comment (handrews) [reply versions/3.1.1.md:2351]: @ralfhandl please feel free to file an issue for that and add it to the "Automation & Infrastructure" GitHub project.  I'd prefer that we not hold up this PR based on the references appendx, as the concerns are orthogonal.

Comment (handrews) [reply versions/3.1.1.md:2351]: Technically, only the markdown is normative, so our normative representation doesn't have that appendix at all.  Therefore it should not block anything.

Comment (ralfhandl) [reply versions/3.1.1.md:2351]: #3740

File: versions/3.1.1.md
Comment (ralfhandl) [inline versions/3.1.1.md:2351]: Does Swagger UI support these already?

Comment (handrews) [reply versions/3.1.1.md:2351]: No, why should that matter?  Sswagger is the slowest-moving major tooling vendor out there.

Comment (handrews) [reply versions/3.1.1.md:2351]: Actually I'm just assuming "no" - the whole point of this change is to raise awareness that this works – please see the _six different issues requesting this feature_ linked in the issue for this PR.  I'll list them here for convenience- I closed them becauase it's technically possible, but got feedback that it's not obvious enough:

* #519 
* #555 
* #773 
* #957 
* #1749 
* #2453

**Comment by handrews** [comment]
To add a little more context (with apologies for the rapid-fire comment spam):

* This feature is (as noted above) highly desired and has been requested repeatedly since 2015
* It is implemented in some places (I believe that @gregsdennis , who wrote the JSON Schema Blog's post about it, supports it in C#)
* `$dynamicRef` and `$dynamicAnchor` are kind of the `discriminator` of 3.1 – there was a long period where I noticed many OpenAPI tools claimed they supported everything, but with an asterisk saying "except `discriminator`... part of the reason for this is that these new keywords are moderately complex.  But part of it is that people don't understand that they can support a common programming paradigm that is otherwise unsupportable.  We will not see these implemented unless they are connected to real customer use cases, and that's what this PR does.

File: versions/3.1.1.md
Comment (ralfhandl) [reply versions/3.1.1.md:2351]: My experience is that what doesn't show up in Swagger UI isn't written into an OpenAPI description. Which shouldn't prevent us from advertising new features 😄

File: versions/3.1.1.md
Comment (ralfhandl) [inline versions/3.1.1.md:2768]:
```suggestion
    genericArrayComponent:
```

File: versions/3.1.1.md
Comment (ralfhandl) [inline versions/3.1.1.md:2793]:
```suggestion
      required: 
      - dataType 
      - data
```
I'd avoid mixing styles

File: versions/3.1.1.md
Comment (ralfhandl) [inline versions/3.1.1.md:2803]:
```suggestion
      - properties:
          dataType:
            const: string
          data:
            $ref: array_of_strings
      - properties:
          dataType:
            const: number
          data:
            $ref: array_of_numbers
```

File: versions/3.1.1.md
Comment (handrews) [reply versions/3.1.1.md:2351]: @ralfhandl I'm glad you're not seeing it as a barrier :-)  But in all seriousness, there are other tooling vendors, and they do have actual customers/users.  I really don't think it's a good idea for the steering committee of a vendor-neutral project to publicly promote the idea that one company's single product is the gating factor for the entire ecosystem.  We've been working very hard for years now to separate "OpenAPI" from "Swagger", and comments like this do the opposite.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

**Review by lornajane — APPROVED** [review]

**Pull request merged**
